### DISCUSSION TITLE
Single Bunch Mode at the Bunch Level

### DISCUSSION THREAD

**Discussion by ejf071189** [ttscoff/bunch-docs#204] [Q&A]
I haven't found a way via frontmatter or commands to make a given bunch close all other active bunches on open when not in single bunch mode. Is this possible?

**Comment by ttscoff** [comment]
No, you can't enable Single Bunch Mode for a single bunch. You can 
manually close other Bunches, though, with `!Name.bunch`.

**Reply by ejf071189** [reply]
Ok that was what I thought. Do you have any recommendation for the quickest way to "close all active bunches". I was thinking of just making a shortcut that filters the bunch folder for .bunch files and for each name, uses the bunch url scheme to close every bunch (can't think of a way to prevent it from running on already closed bunches but hopefully that wouldn't slow things down too much). Other concern might be odd behavior if the bunches close in parallel rather than sequentially (file to file I mean, all of my bunches individually run in sequential mode currently)

**Comment by ttscoff** [comment]
If you're constantly adding and removing Bunches, yeah, a script that 
loops through all .bunch files would be the simplest, most maintainable 
solution. You might find that using writing it in AppleScript to loop 
through the directory and calling `close bunch "bunch name"` (in a `tell 
app "Bunch"` block) instead of the URL handler might be more predictable 
for sequencing.

If you have a pretty static collection of Bunches, the superior solution 
would be to just list them in the Bunch that should close them. This 
will do a better job of tracking what apps are open, allow better 
sequencing, and will avoid quitting any apps from other Bunches that the 
current Bunch is going to open anyway. If this is something you need to 
repeat between multiple Bunches, you could make it a snippet and just 
maintain the list in one place.

```
!Comms.bunch
!Work.bunch
!Game Time.bunch
```

If you're handy with scripting, you could pretty easily generate this 
list automatically and just run the script to update it when needed (or 
have it run as a "run before" script in the Bunch, then use a snippet 
call to the generated file). You could also use a script variable that 
populated itself with the above content at runtime and then executed as 
a snippet..., but I'd have to test to see if that worked or not.

All of this said, I can see room for a `(close every Bunch)` command 
that would simplify this process. Already have one that quits all apps, 
so why not? I could also see modifying the `single bunch mode:` key to 
allow a value of `true`, which would cause opening that Bunch to close 
other Bunches, regardless of the Single Bunch Mode preference. It's just 
that it would then have a different meaning than "single bunch mode: 
ignore" which affects how it reacts to other Bunches opening, not how it 
opens itself. Might need a different key for that behavior, like `close 
other: true` or something. Will have to think about it.

-Brett

**Reply by ejf071189** [reply]
Appreciate such a thoughtful response! I'm still fairly early in my bunch design process so the maintained list of bunches to close in a snippet *should* be a good solution, but I've been thinking about making quite a long list of bunches for different projects. Using variables and interactive prompts should be able to condense the list and make creation of new bunches infrequent enough.

Besides the script maintenance side, the attractive thing about "single bunch mode" is that it simplifies context shifting since you don't have to close the current bunch before opening a new one. I'm really drawn to that concept but I also like the flexibility of having "ad hoc bunch modules" that I can open and close on the fly. In a perfect world, I'd be able to set a "bunch group" key that would tie a group of bunches with the same value into a radio button in which only one bunch in the group can be active at any time, and then all ungrouped bunches would be independent check boxes. I'm not sure if that would be harder to implement than the "close all bunches" idea but I think it would be even more powerful in being able to combine the efficiency of single bunch mode with the flexibility of simultaneous bunches.

**Comment by ttscoff** [comment]
It won’t solve that last feature request, but not that you can use tags to open and close groups of bunches.

https://bunchapp.co/docs/bunch-files/tags/

- Brett

**Reply by ejf071189** [reply]
Ah I hadn't seen that one; I think that actually gives me all the functionality I'm looking for! The radio/checkbox system would certainly make it more clear to the user, but as my own workaround just prefixing the name with a group or descriptor should work just fine.

**Reply by ejf071189** [reply]
I think I'm actually impeded by this:
> Be aware that if you call a \tag command from within a Bunch containing that tag, that Bunch will also be affected. Be cautious as results can be unpredictable if Bunches calling Bunches get nested too deeply.

I seem to be gettin getting erratic behavior whenever I close a tag within a like-tagged bunch whether I use !\tag or %!\tag

**Comment by ttscoff** [comment]
Yeah, that feature is a bit underdeveloped as I wasn’t sure yet that there was a solid use case for it :). It’s perfect for this kind of application, though (if it works), so I’ll spend some time debugging that this week if I’m able.

- Brett

**Reply by ejf071189** [reply]
I actually realized that (assuming I'm not making a mistake) %! commands don't seem to work at all! My expected behavior was that for a "child.bunch", that `%!child.bunch` would close the bunch and then not reopen the child bunch when the parent closes; whereas as far as I can tell, the line is just ignored altogether (I'm guessing there's no implementation for combining the two prefixes). I'm not sure if implementing that would resolve the issue more easily than modifying the \tag behavior, but thought I'd mention it!

**Reply by ejf071189** [reply]
One more thing I noticed is that the tags seem to be cascading to bunches that are opened within another bunch. So if you have
```
---
title: Bunch1
tags: tag1
---
Google Chrome
Bunch2.bunch
```
Bunch2 will show in the app menu as being tagged with tag1, even if it doesn't have it in its own frontmatter.

**Comment by ttscoff** [comment]
I think I have both issues solved, will release a beta version today for 
testing. I was forgetting entirely to check for the "ignored" when 
handling tags.

Bunches launched by another Bunch inherit the parent's frontmatter, 
which has its uses, but certain keys get removed and not passed on. I 
can't think of a reason you'd want tags to be inherited, so I'm adding 
that to the removed keys. It should solve the issue you're seeing, we'll 
see if it causes any other issues in the process.

-Brett

**Comment by ttscoff** [comment]
Ok, I released beta 1.4.8, please test and let me know if it resolves 
the tagging issues or not. https://bunchapp.co/download/

**Reply by ejf071189** [reply]
The cascading tags and not following the % (ignore) seem fixed. I'm not sure if this was intended to fix the issue of not being able to close tags from a like-tagged bunch, but that appears to still be not working when multiple tagged bunches start with the same closing command (does work when just one bunch in the group has the command). Test config and cases below.

## Bunch config:
```
---
title: tag1 bunch 1
tags: tag1
---
// test below with both "!" and "%!" prefix as well as commented out
!\tag1
Mail

-----------------------------

---
title: tag1 bunch 2
tags: tag1
---
// test next two lines both with "!" and "%!" prefix
!\tag1
!\tag2
test-untagged.bunch
TextEdit

-----------------------------

---
title: tag2
tags: tag2
---
Calculator

-----------------------------

---
title: untagged
---
Calendar

```

## no ignores (!)
- Open tag1 bunch 1
	- App state
		- Mail is opened
	- Bunch state
		- tag1 bunch 1 opened
- Open tag2
	- App state
		- Mail remained open
		- Calculator opened
	- Bunch state
		- tag1 bunch 1 remained open
		- tag2 opened
- Open tag1 bunch 2
	- App state
		- Calculator closed
		- Mail remained open (FAILURE: should have closed)
		- Calendar opened (from untagged bunch)
		- TextEdit opened (from tag1 bunch 2)
	- Bunch state
		- tag1 bunch 2 opened
		- tag1 bunch 1 remained open (FAILURE: should have closed)
		- tag2 closed
		- untagged opened
- Close tag1 bunch 2
	- App state
		- Mail remained open
		- Calculator reopened
		- Calendar closed
		- TextEdit closed
	- Bunch state
		- tag1 bunch 1 remained open
		- tag2 reopened
		- untagged closed
		- tag1 bunch 2 closed

## ignores (%!)
- Open tag1 bunch 1
	- Same as previous
- Open tag2
	- Same as previous
- Open tag1 bunch 2
	- Same as previous (including FAILURE)
- Close tag1 bunch 2
	- App state
		- Mail remained open
		- Calculator did not reopen
		- Calendar closed
		- TextEdit closed
	- Bunch state
		- tag1 bunch 1 remained open
		- tag2 did not reopen
		- untagged closed
		- tag1 bunch 2 closed
## comment out the `!\tag1` line of "tag1 bunch 1"
- Open tag1 bunch 1
	- Same as previous
- Open tag2
	- Same as previous
- Open tag1 bunch 2
	- "tag1 bunch 1" and mail app closed as expected
- Close tag1 bunch 2
	- same as previous

**Comment by ejf071189** [comment]
Ok that was what I thought. Do you have any recommendation for the quickest
way to "close all active bunches". I was thinking of just making a shortcut
that filters the bunch folder for .bunch files and for each name, uses the
bunch url scheme to close every bunch (can't think of a way to prevent it
from running on already closed bunches but hopefully that wouldn't slow
things down too much).

**Comment by ttscoff** [comment]
AppleScript is the most effective way to do it. See [list all open 
bunches](https://bunchapp.co/docs/integration/applescript/#list-all-bunches) 
and [close 
bunch](https://bunchapp.co/docs/integration/applescript/#list-all-bunches).

Essentially you would do (not tested, from top of mind):

```
tell application "Bunch"
	repeat with _bunch in (list open bunches)
		close bunch _bunch
	end repeat
end tell
```

-Brett
